packages/store-sync/src/recs/syncToRecs.ts

```diff
@@ -0,0 +1,227 @@
+  destroy: () => void;
+};
+
+export async function syncToRecs<
+  TConfig extends StoreConfig = StoreConfig,
+  TComponents extends Record<string, RecsComponent<RecsSchema, StoreComponentMetadata>> = Record<
+    string,
+    RecsComponent<RecsSchema, StoreComponentMetadata>
+  >
+>({
+  world,
+  config,
+  address,
+  publicClient,
+  components: initialComponents,
+  initialState,
+}: SyncToRecsOptions<TConfig, TComponents>): Promise<SyncToRecsResult<TConfig, TComponents>> {
+  const components = {
+    ...initialComponents,
+    ...defineInternalComponents(world),
+  };
+
+  const singletonEntity = world.registerEntity({ id: hexKeyTupleToEntity([]) });
+
+  let startBlock = 0n;
+
+  if (initialState != null && initialState.blockNumber != null) {
+    debug("hydrating from initial state to block", initialState.blockNumber);
+    startBlock = initialState.blockNumber + 1n;
+
+    setComponent(components.SyncProgress, singletonEntity, {
+      step: SyncStep.SNAPSHOT,
+      message: `Hydrating from snapshot to block ${initialState.blockNumber}`,
+      percentage: 0,
+    });
+
+    const componentList = Object.values(components);
+
+    const numRecords = initialState.tables.reduce((sum, table) => sum + table.records.length, 0);
+    let recordsProcessed = 0;
+
+    for (const table of initialState.tables) {
+      setComponent(components.TableMetadata, getTableKey(table) as Entity, { table });
+      const component = componentList.find((component) => component.id === table.tableId);
+      if (component == null) {
+        debug(`no component found for table ${table.namespace}:${table.name}, skipping initial state`);
+        continue;
+      }
+      for (const record of table.records) {
+        const entity = encodeEntity(table.keySchema, record.key);
+        setComponent(component, entity, record.value as ComponentValue);
+
+        recordsProcessed++;
+        setComponent(components.SyncProgress, singletonEntity, {
+          step: SyncStep.SNAPSHOT,
+          message: `Hydrating from snapshot to block ${initialState.blockNumber}`,
+          percentage: (recordsProcessed / numRecords) * 100,
+        });
+      }
+      debug(`hydrated ${table.records.length} records for table ${table.namespace}:${table.name}`);
+    }
+  }
+
+  // TODO: if startBlock is still 0, find via deploy event
+
+  debug("starting sync from block", startBlock);
+
+  const latestBlock$ = createBlockStream({ publicClient, blockTag: "latest" }).pipe(share());
+
+  const latestBlockNumber$ = latestBlock$.pipe(
+    filter(isNonPendingBlock),
+    map((block) => block.number),
+    share()
+  );
+
+  const latestBlockNumber = new BehaviorSubject<bigint | null>(null);
+  {
+    const sub = latestBlockNumber$.subscribe(latestBlockNumber);
+    world.registerDisposer(() => sub.unsubscribe());
+  }
+
+  const blockLogs$ = latestBlockNumber$.pipe(
+    tap((blockNumber) => {
+      debug("latest block number", blockNumber);
+    }),
```

Comment:
why does `latestBlockNumber` need to be a `BehaviorSubject`? Could we just update a simple variable in this tap? (Would also save us from having to manage the subscription / registering the disposer in the World)

Reply:
I originally had wired up like that, but it felt a little dirty. I think the BehaviorSubject approach is effectively the same thing, minus the additional subscriber, and felt cleaner to me.

Reply:
Can you elaborate on what was harder to read / felt dirtier?

To me it seems like the BehaviorSubject is just unnecessary overhead since we only ever read the current value via `.value`. (Like it just feels weird to have all this code for a variable https://github.com/ReactiveX/rxjs/blob/master/src/internal/BehaviorSubject.ts, https://github.com/ReactiveX/rxjs/blob/master/src/internal/Subject.ts, https://github.com/ReactiveX/rxjs/blob/master/src/internal/Observable.ts even if there is little overhead at runtime). The other thing that feels a bit weird is that we're breaking the convention of using the `$` suffix for streams here.

For the point of easier to read, we could still update the `latestBlockNumber` as a variable in a subscription like here instead of in `tap`.

Reply:
chatted in person and ended up moving off of behavior subject